planning/mission_planner/src/mission_planner_interface.cpp

```diff
@@ -0,0 +1,148 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mission_planner/mission_planner_interface.hpp"
+
+#include <lanelet2_extension/utility/message_conversion.hpp>
+#include <lanelet2_extension/utility/query.hpp>
+#include <lanelet2_extension/visualization/visualization.hpp>
+
+#include <lanelet2_routing/Route.h>
+#ifdef ROS_DISTRO_GALACTIC
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+#else
+#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
+#endif
+#include <visualization_msgs/msg/marker_array.h>
+
+#include <memory>
+#include <string>
+
+namespace mission_planner
+{
+MissionPlannerInterface::MissionPlannerInterface(
+  const std::string & node_name, const rclcpp::NodeOptions & node_options)
+: Node(node_name, node_options), tf_buffer_(get_clock()), tf_listener_(tf_buffer_)
+{
+  map_frame_ = declare_parameter("map_frame", "map");
+
+  using std::placeholders::_1;
+
+  odometry_subscriber_ = create_subscription<nav_msgs::msg::Odometry>(
+    "/localization/kinematic_state", rclcpp::QoS{1},
+    std::bind(&MissionPlannerInterface::odometry_callback, this, _1));
+  goal_subscriber_ = create_subscription<geometry_msgs::msg::PoseStamped>(
+    "input/goal_pose", 10, std::bind(&MissionPlannerInterface::goal_pose_callback, this, _1));
+  check_point_subscriber_ = create_subscription<geometry_msgs::msg::PoseStamped>(
+    "input/checkpoint", 10, std::bind(&MissionPlannerInterface::check_point_callback, this, _1));
+
+  rclcpp::QoS durable_qos{1};
+  durable_qos.transient_local();
+  route_publisher_ =
+    create_publisher<autoware_auto_planning_msgs::msg::HADMapRoute>("output/route", durable_qos);
+  marker_publisher_ =
+    create_publisher<visualization_msgs::msg::MarkerArray>("debug/route_marker", durable_qos);
+}
+
+void MissionPlannerInterface::odometry_callback(const nav_msgs::msg::Odometry::ConstSharedPtr msg)
+{
+  ego_pose_ = std::make_shared<const geometry_msgs::msg::Pose>(msg->pose.pose);
+}
+
+boost::optional<geometry_msgs::msg::PoseStamped> MissionPlannerInterface::transform_pose(
+  const geometry_msgs::msg::PoseStamped & input_pose, const std::string & target_frame)
+{
+  try {
+    geometry_msgs::msg::PoseStamped output_pose;
+    const auto transform =
+      tf_buffer_.lookupTransform(target_frame, input_pose.header.frame_id, tf2::TimePointZero);
+    tf2::doTransform(input_pose, output_pose, transform);
+    return output_pose;
+  } catch (tf2::TransformException & ex) {
+    RCLCPP_WARN(get_logger(), "%s", ex.what());
+  }
+
+  return {};
+}
+
+void MissionPlannerInterface::goal_pose_callback(
+  const geometry_msgs::msg::PoseStamped::ConstSharedPtr goal_msg_ptr)
+{
+  // set start pose
+  if (!ego_pose_) {
+    RCLCPP_ERROR(get_logger(), "Ego pose has not been subscribed. Aborting mission planning");
+    return;
+  }
```

Comment:
Should there be transformation to `map_frame_` after this validation just in case?
Although it would be a rare case, since we are making map_frame_ configurable, we never know in which frame user would like to calculate route.

Something similar to goal_pose:
```
  const auto opt_transformed_ego_pose = transform_pose(ego_pose_, map_frame_);
  if (!opt_transformed_ego_pose) {
    RCLCPP_ERROR(get_logger(), "Failed to get vehicle pose in map frame. Aborting mission planning");
    return;
  }
  const auto transformed_ego_pose = opt_transformed_ego_pose.get();
```

Reply:
Ah, right.
I guess we have to apply the same transformation to ego pose in other planning modules. What do you think?

Reply:
Yes, I think that would be a good idea. If it makes sense to do that as separate PR, I can approve this PR as is for now.

Reply:
Then P/C team and I will work on it with another PR. Could you approve please?